From fc426834088c27978113d544f12acf685780972a Mon Sep 17 00:00:00 2001
From: Xie XiuQi <xiexiuqi@huawei.com>
Date: Fri, 26 Feb 2021 20:21:28 +0800
Subject: [PATCH] arm64/mpam: unmap all previous address when failed
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: fc426834088c27978113d544f12acf685780972a


hulk inclusion
category: bugfix
bugzilla: 48265
CVE: NA

--------------------------------

Signed-off-by: Xie XiuQi <xiexiuqi@huawei.com>
Reviewed-by: Hanjun Guo <guohanjun@huawei.com>
Signed-off-by: Yang Yingliang <yangyingliang@huawei.com>
Signed-off-by: Wang ShaoBo <bobo.shaobowang@huawei.com>
Reviewed-by: Cheng Jian <cj.chengjian@huawei.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 arch/arm64/kernel/mpam.c | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/arch/arm64/kernel/mpam.c b/arch/arm64/kernel/mpam.c
index d0b4f6ab70ad..a83b43c9dcaa 100644
--- a/arch/arm64/kernel/mpam.c
+++ b/arch/arm64/kernel/mpam.c
@@ -101,6 +101,21 @@ struct mpam_node mpam_node_all[] = {
 	MPAM_NODE(HHAALL3, MPAM_RESOURCE_MC, 0x200090ULL, 0xC1),
 };
 
+void mpam_nodes_unmap(void)
+{
+	int i;
+	size_t num_nodes = ARRAY_SIZE(mpam_node_all);
+	struct mpam_node *n;
+
+	for (i = 0; i < num_nodes; i++) {
+		n = &mpam_node_all[i];
+		if (n->base) {
+			iounmap(n->base);
+			n->base = NULL;
+		}
+	}
+}
+
 int mpam_nodes_init(void)
 {
 	int i, ret = 0;
@@ -111,6 +126,10 @@ int mpam_nodes_init(void)
 		n = &mpam_node_all[i];
 		ret |= cpulist_parse(n->cpus_list, &n->cpu_mask);
 		n->base = ioremap(n->addr, 0x10000);
+		if (!n->base) {
+			mpam_nodes_unmap();
+			return -ENOMEM;
+		}
 	}
 
 	return ret;
-- 
2.26.2

